File: web/gi-kg-viewer/src/utils/operatorYamlProfile.ts
/** Split / merge top-level ``profile:`` for operator YAML (no full YAML parser). */

/** Quote a scalar for a one-line ``profile:`` value when needed (safe YAML subset). */
export function yamlScalarForProfileLine(name: string): string {
  const s = name.trim()
  if (!s) {
    return '""'
  }
  if (/^[A-Za-z0-9_.-]+$/.test(s)) {
    return s
  }
  return JSON.stringify(s)
}

export function splitOperatorYamlProfile(content: string): { profile: string; body: string } {
  const raw = content.replace(/\r\n/g, '\n')
  const lines = raw.split('\n')
  for (let i = 0; i < lines.length; i += 1) {
    const m = lines[i].match(/^\s*profile:\s*(.+?)\s*$/)
    if (m) {
      let v = m[1].trim()
      const hash = v.indexOf('#')
      if (hash >= 0) {
        v = v.slice(0, hash).trim()
      }
      if (
        (v.startsWith('"') && v.endsWith('"')) ||
        (v.startsWith("'") && v.endsWith("'"))
      ) {
        v = v.slice(1, -1)
      }
      const rest = [...lines.slice(0, i), ...lines.slice(i + 1)].join('\n')
      return { profile: v, body: rest.replace(/^\n+/, '').replace(/\n+$/, '') }
    }
  }
  return { profile: '', body: raw.replace(/\n+$/, '') }
}

export function mergeOperatorYamlProfile(profile: string, body: string): string {
  const b = body.replace(/^\n+/, '').replace(/\n+$/, '')
  if (!profile.trim()) {
    return b ? `${b}\n` : ''
  }
  const p = yamlScalarForProfileLine(profile.trim())
  if (!b) {
    return `profile: ${p}\n`
  }
  return `profile: ${p}\n${b}\n`
}

File: src/podcast_scraper/server/operator_yaml_profile.py
"""Split / merge top-level ``profile:`` for viewer operator YAML.

Mirrors SPA ``operatorYamlProfile.ts``.
"""

from __future__ import annotations

import json
import re
from pathlib import Path


def format_profile_scalar(name: str) -> str:
    """Quote a scalar for a one-line ``profile:`` value when needed (safe YAML subset)."""
    s = name.strip()
    if not s:
        return '""'
    if re.match(r"^[A-Za-z0-9_.-]+$", s):
        return s
    return json.dumps(s)


def split_operator_yaml_profile(content: str) -> tuple[str, str]:
    """Return ``(profile_name, body_without_profile_line)``."""
    raw = content.replace("\r\n", "\n")
    lines = raw.split("\n")
    for i, line in enumerate(lines):
        m = re.match(r"^\s*profile:\s*(.+?)\s*$", line)
        if not m:
            continue
        v = m.group(1).strip()
        h = v.find("#")
        if h >= 0:
            v = v[:h].strip()
        if (v.startswith('"') and v.endswith('"')) or (v.startswith("'") and v.endswith("'")):
            v = v[1:-1]
        rest = "\n".join(lines[:i] + lines[i + 1 :])
        body = rest.lstrip("\n").rstrip("\n")
        return v, body
    return "", raw.rstrip("\n")


def _body_has_operator_substance(body: str) -> bool:
    """True when ``body`` has any non-blank, non-comment line."""
    for line in body.replace("\r\n", "\n").split("\n"):
        s = line.strip()
        if not s or s.startswith("#"):
            continue
        return True
    return False


def merge_operator_yaml_profile(profile: str, body: str) -> str:
    """Emit YAML with optional first-line ``profile:`` then ``body``."""
    b = body.lstrip("\n").rstrip("\n")
    if not profile.strip():
        return f"{b}\n" if b else ""
    p = format_profile_scalar(profile.strip())
    if not b:
        return f"profile: {p}\n"
    return f"profile: {p}\n{b}\n"


def expand_profile_only_with_packaged_example(
    content: str,
    *,
    example_path: Path | None,
) -> str:
    """If ``content`` has no keys beyond ``profile:``/comments, append example body."""
    if example_path is None or not example_path.is_file():
        return content
    profile, body = split_operator_yaml_profile(content)
    if _body_has_operator_substance(body):
        return content
    example_text = example_path.read_text(encoding="utf-8", errors="replace")
    _, ex_body = split_operator_yaml_profile(example_text)
    if not _body_has_operator_substance(ex_body):
        return content
    ex_body = ex_body.strip() + "\n"
    if profile.strip():
        return f"profile: {format_profile_scalar(profile)}\n{ex_body}"
    return ex_body
